### ISSUE TITLE
Integrate Intel TBB

### ISSUE THREAD

**Issue by wds15** [stan-dev/math#1377]
## Description
The Intel TBB needs integration with the stan-math library in a few areas:

- threadpool initialization based on `STAN_NUM_THREADS` environment variable 
- initialization of the AD tape resource of reverse mode for threads entering the threadpool
- switch of `map_rect` use of C++11 threads over to use the TBB threadpool

## Example

Nothing changes for end-users.

Interfaces will have to call the threadpool initialization function - otherwise the TBB default will be used (which is to use all cores).

## Expected Output

Things should run faster and developers can deal with coding up high-level tasks rather than low-level threads.

#### Current Version:
v2.20.0

**Comment by alashworth** [comment]
Which unit tests are affected by this change? With the recently merged pull request, what stan threading tests should be run with and without TBB support?

**Comment by rok-cesnovar** [comment]
The recently merged pull request actually changes nothing to be exact. For now it just links the TBB, and also links tbbmalloc on Mac. No tests have been altered so far.

TBB will start actually being used when https://github.com/stan-dev/math/pull/1376 is merged, which will probably happen today, if there will be no test hiccups or requested changes.

TBB will then be used for all `map_rect` with STAN_THREADS tests.

The PR itself only touches:
- test/unit/math/rev/mat/functor/gradient_test.cpp

Creates the following new tests:
- test/unit/math/prim/core/get_num_threads_test.cpp
- test/unit/math/prim/core/init_threadpool_tbb_late_test.cpp
- test/unit/math/prim/core/init_threadpool_tbb_test.cpp

And 1 test is deleted:

- test/unit/math/prim/mat/functor/num_threads_test.cpp will be deleted

But to reiterate so you wont waste any time today, none of this is currently on develop.

**Comment by wds15** [comment]
The answer is simpler than that... the TBB is mandatory for everything. The code which setups up the AD tape is using TBB bits and pieces. The TBB supports *optional* parallelism such that code can be written with the TBB which works in single-core and multi-core mode.

**Comment by alashworth** [comment]
Is STAN_THREADS going to be the default configuration of math/cmdstan and friends in the future?

**Comment by wds15** [comment]
Not yet - but we will get to the all-thread future almost for sure. That will happen when we have threaded warmup in Stan. Since this almost certainly has significant speedups for the entire user base, everyone will want to use that... but we are not there yet and right now this is my judgement call.

Anyway, the TBB will be used regardless of STAN_THREADS (recall, the TBB supports threaded and non-threaded programs). Doing so allows us to have a cleaner codebase.

As of now there is no TBB code in Stan-math; we just updated the makefiles. Later today when #1376 is merged, then the TBB is required.
